### DISCUSSION TITLE
'EffectiveStrain' in 'RankTwoScalarTools'

### DISCUSSION THREAD

**Discussion by rh201** [idaholab/moose#23766] [Q&A Modules: Solid mechanics] [answered]
Hi MOOSEer,

I try to calculate/use the equivalent plastic strain as shows https://mooseframework.inl.gov/source/utils/RankTwoScalarTools.html
and discussed here #[20979](https://github.com/idaholab/moose/discussions/20979) and the input as
> [Auxvariables]
  [./equivalent_plastic_strain]
    order = CONSTANT
    family = MONOMIAL
  [../]
[]
[Auxkernels]
  [./equivalent_plastic_strain]
    type =  RankTwoScalarTools #RankTwoScalarAux 
    rank_two_tensor =  plastic_strain
    variable = equivalent_plastic_strain
    scalar_type = effectiveStrain
  [../]
[]

But the error throw out as

> *** ERROR ***
A 'RankTwoScalarTools' is not a registered object.

Maybe I misunderstood it, any suggestions are highly appreciated. Thanks.

**Comment by GiudGiud** [comment, marked as answer]
Hello

Utils are meant to be called from the code, not from the input file. 
I think having an auxkernel was more appropriate there, to fill an auxvariable

Guillaume

**Reply by rh201** [reply]
Thanks. I don't quite get it, could you explain more in detail? or are there related examples?

**Reply by GiudGiud** [reply]
You’re trying to use a utility. 
utilities are not objects you create in input files. It s code you can use when you are developing

**Reply by rh201** [reply]
Yes. I'm not sure where should I call for utility. Do you mean a new Auxkernel code containing RankTwoScalarTools and using utility needs to be developed?

**Reply by GiudGiud** [reply]
yeah you could develop a new auxkernel that uses the utility you want to use. 
What are you trying to do?

I think you just want to use the RankTwoScalarAux auxkernel here

**Reply by rh201** [reply]
I want to call for the equivalent/effective plastic strain for other material modulus. 

so to put it this way, how can I obtain the \epsilon_p^eq here? Thanks
![image](https://user-images.githubusercontent.com/79764031/225876042-c5dbcc37-0cd7-4e6c-aac7-b46fb05fcf64.png)

**Reply by GiudGiud** [reply]
If you add 
```
[Debug]
  show_material_props= true
[]
```
to the simulation do you see that property declared?

Guillaume

**Reply by rh201** [reply]
Hi,
1. I add it and it shows as

> Consumed Material Properties:
      Object: AuxKernel::isotropic_plasticityeffective_plastic_strain
  Properties: effective_plastic_strain
      Object: AuxKernel::isotropic_plasticityhardening_variable
  Properties: hardening_variable

so it means it's calculated, right? 

2. Previously, I used the [Compute Multiple Inelastic Stress] and [Isotropic Plasticity Stress Update] and input as follows to retrieve  \epsilon_p^eq, but is this  \epsilon_p^eq here the same as the one I showed in the figure? If yes, then what's the difference of [ effectiveStrain method] in https://mooseframework.inl.gov/source/utils/RankTwoScalarTools.html, as discussed as #20979. If no, then what's the definition of the \epsilon_p^eq in the two modules.  
>    [./effective_plastic_strain]
    type = MaterialRealAux
    property =  effective_plastic_strain
    variable =  effective_plastic_strain
  [../]
 
3. the type of \epsilon_p^eq should be  MaterialRealAux or RankTwoScalarAux or RankTwoScalarTools? **All My questions come down to how can I calculate and call for \epsilon_p^eq.** Thanks

**Reply by GiudGiud** [reply]
@hugary1995 @jiangwen84

**Reply by rh201** [reply]
Additionally, I'm trying to run the original input file from 'modules/combined/test/tests/j2_plasticity_vs_LSH/j2_hard1_mod_small_strain.i', which use the auxkernel to get eq_pl_strain as

>   [./eq_pl_strain]
    type = RankTwoScalarAux
    rank_two_tensor = plastic_strain
    scalar_type = EffectiveStrain
    variable = eq_pl_strain
  [../]

 but the error is still there. 

> *** ERROR ***
To compute an effective inelastic strain use RankTwoScalarTools::effectiveStrain()

**Reply by hugary1995** [reply]
```python
[AuxVariables]
  [ep]
    order = CONSTANT
    family = MONOMIAL
    [AuxKernel]
      type = MaterialRealAux
      property = effective_plastic_strain
      execute_on = 'TIMESTEP_END'
    []
  []
[]
```
